### PR TITLE
[GH-1102] retry on 503

### PR DESCRIPTION
**Pull request by ehigham**
### Purpose

https://broadinstitute.atlassian.net/browse/GH-1102

Google has been giving us lots of 503s, resulting in partial uploads and PTC failures.
In the case of getting these errors, we should retry.

I don't know how to test this.

### CODE REVIEW

File: src/ptc/util/misc.clj [hunk 1]
```diff
@@ -83,9 +86,7 @@
   "Run ARGS in a shell and return stdout or throw."
   [& args]
   (let [{:keys [exit err out]} (apply shell/sh args)]
-    (when-not (zero? exit)
-      (throw (Exception. (format "%s: %s exit status from: %s : %s"
-                                 ptc/the-name exit args err))))
+    (when-not (zero? exit) (throw (IOException. err)))
```
Comment (tbl3rd) [inline src/ptc/util/misc.clj:88]: Yay.

**Review by jack-r-warren — APPROVED** [review]
Agree on it being impossible to test, LGTM

File: src/ptc/util/misc.clj [hunk 2]
```diff
@@ -10,7 +10,9 @@
             [vault.client.http]         ; vault.core needs this
             [vault.core            :as vault])
   (:import [java.util UUID]
-           [org.apache.commons.mail SimpleEmail]))
+           [java.util.concurrent TimeUnit]
+           [org.apache.commons.mail SimpleEmail]
```
Comment (tbl3rd) [inline src/ptc/util/misc.clj:14]: Does PTC mail?

File: src/ptc/util/misc.clj [hunk 3]
```diff
@@ -144,17 +144,32 @@
...
+      (or (try
+            (apply shell! "gsutil" args)
+            (catch IOException ex
+              (when-not (and (str/includes? (.getMessage ex) "503 Server Error")
+                             (< attempt max))
+                (throw ex))
+              (log/warnf "received 503 (attempt % of %s)" attempt max)
+              (log/info "sleeping before another attempt")
+              (.sleep TimeUnit/SECONDS 30)))
+          (recur (inc attempt))))))
```
Comment (tbl3rd) [inline src/ptc/util/misc.clj:160]: Awesome code.

File: test/ptc/tools/gcs.clj [hunk 4]
```diff
@@ -76,39 +76,23 @@
...
-   (http/request {:method  :delete ;; :debug true :debug-body true
-                  :url     (bucket-object-url bucket object)
-                  :headers headers}))
-  ([bucket object]
-   (delete-object bucket object (get-auth-header!)))
   ([url]
-   (apply delete-object (parse-gs-url url))))
+   (misc/gsutil "rm" url))
+  ([bucket object]
+   (delete-object (str "gs://" bucket "/" object))))
```
Comment (tbl3rd) [inline test/ptc/tools/gcs.clj:88]: Ah. OK.

**Review by tbl3rd — APPROVED** [review]

File: src/ptc/util/misc.clj (hunk 2 above)
Comment (ehigham) [reply src/ptc/util/misc.clj:14]: I don't know. Not fixing here.

File: src/ptc/util/misc.clj (hunk 3 above)
Comment (ehigham) [reply src/ptc/util/misc.clj:160]: How modest of you.

**Pull request merged**